### PR TITLE
test(menu): refactor for cypress tests

### PR DESCRIPTION
**Pull request by stephenogorman**
### Proposed behaviour

Refactor Cypress tests for the `Menu` component to use new cypress-component-react framework for testing.
Add `menu.test.js`. 

### Current behaviour

Current process is the regular Cypress approach with BDD.

### Checklist

- [x] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required

#### QA

- [x] Tested in CodeSandbox/storybook
- [x] Add new Cypress test coverage if required
- [x] Carbon implementation matches Design System/designs
- [x] UI Tests GitHub check reviewed if required

### Testing instructions

- [x] Run `npx cypress open --component` to check if there is newly added test.file
- [x] Check if the `menu.test.js` file passed
- [x] Check all tests were moved/refactored from .feature file -> .test.js file
- [x] Check all unnecessary files were deleted, including .json file
- [x] Run `npx cypress run --component` to check none of the other *.test.js files have regressed
- [x] Run `npx cypress run --e2e` to check none of the feature files have regressed

### CODE REVIEW

File: src/components/menu/menu.test.js [hunk 1]
```diff
@@ -0,0 +1,1260 @@
...
+        getComponent("menu-fullscreen").should(
+          "have.css",
+          "right",
+          `${right}px`
+        );
+        menuItem().eq(positionOfElement("first"), div).click();
+        getComponent("menu-fullscreen").should(
+          "have.css",
+          "left",
+          `${after}px`
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:606]: do we need this `after` if it is still `0` for both examples?

File: src/components/menu/menu.test.js [hunk 2]
```diff
@@ -0,0 +1,1260 @@
...
+          "right",
+          `${after}px`
+        );
+      }
+    );
+
+    it("should verify that the width of the whole submenu with a long-labelled item is determined by the width of the item", () => {
+      CypressMountWithProviders(
+        <Box mb={150}>
+          {["white"].map((menuType) => (
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:619]: do we need a `.map` here if we are using only one item in array?

File: src/components/menu/menu.test.js [hunk 3]
```diff
@@ -0,0 +1,1260 @@
...
+      submenu().eq(positionOfElement("first")).trigger("mouseover");
+      innerMenu(positionOfElement("second"), span).then(($item) => {
+        submenuBlock().should("have.css", "width", `${$item.width()}px`);
+      });
+    });
+
+    it("should verify that the width of the whole submenu with a long title is determined by the width of the title", () => {
+      CypressMountWithProviders(
+        <Box mb={150}>
+          {["white"].map((menuType) => (
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:645]: same question here about the `.map`

File: src/components/menu/menu.test.js [hunk 4]
```diff
@@ -0,0 +1,1260 @@
...
+
+      submenu().eq(positionOfElement("first")).trigger("mouseover");
+      submenu()
+        .eq(positionOfElement("first"))
+        .then(($menu) => {
+          submenuBlock().should("have.css", "width", `${$menu.width()}px`);
+        });
+    });
+
+    it.each([
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:668]: we can wrap this test into a separate `describe` and pass what exactly `type` we are passing, as I did in line 104 in `pages.test.js` file

File: src/components/menu/menu.test.js [hunk 5]
```diff
@@ -0,0 +1,1260 @@
...
+      [15, 15],
+      [27, 27],
+      [41, 41],
+      ["10px", 10],
+      ["30px", 30],
+      ["50px", 50],
+    ])(
+      "should verify Menu height is set by height prop when passed as an integer or a string",
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:695]: same here for a type, we can actually wrap all these tests where we are passing this sentence `when passed as an integer or a string`

File: src/components/menu/menu.test.js [hunk 6]
```diff
@@ -0,0 +1,1260 @@
...
+    it.each([
+      [10, "10"],
+      [50, "50"],
+      [100, "100"],
+    ])("should verify Menu flex shrink is %s", (value, shrinkText) => {
+      CypressMountWithProviders(
+        <MenuComponent flex="auto" flexShrink={value} />
+      );
+
+      menu("box").should("have.css", "flex-shrink", shrinkText);
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:936]: why here we have `menu("box")` but everywhere else just `menu()` ?

File: src/components/menu/menu.test.js [hunk 7]
```diff
@@ -0,0 +1,1260 @@
...
+
+    it.each([
+      ["default", 1, 16, 116],
+      ["large", 4, 0, 148],
+    ])(
+      "should verify that a Menu size is %s",
+      (size, height, margin, width) => {
+        CypressMountWithProviders(
+          <Box mb={150}>
+            {["white"].map((menuType) => (
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1080]: same here for the `.map(...)`

File: src/components/menu/menu.test.js [hunk 8]
```diff
@@ -0,0 +1,1260 @@
...
+
+    it.each([
+      ["default", "rgba(0, 0, 0, 0)"],
+      ["alternate", "rgb(217, 224, 228)"],
+    ])(
+      "should verify that a Menu Segment Title has a variant background color",
+      (variant, color) => {
+        CypressMountWithProviders(
+          <Box mb={150}>
+            {["white"].map((menuType) => (
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1148]: same here for the `.map(...)`

File: src/components/menu/menu.test.js [hunk 9]
```diff
@@ -0,0 +1,1260 @@
...
+    it("should call onSubmenuOpen callback when a click event is triggered", () => {
+      CypressMountWithProviders(
+        <MenuComponentItems onSubmenuOpen={callback} />
+      );
+
+      menuComponent(positionOfElement("second"))
+        .click()
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1212]:
```suggestion
          expect(callback).to.have.been.calledOnce;
```

File: src/components/menu/menu.test.js [hunk 10]
```diff
@@ -0,0 +1,1260 @@
...
+      (key) => {
+        CypressMountWithProviders(
+          <MenuComponentItems onSubmenuOpen={callback} />
+        );
+
+        menuComponent(positionOfElement("second"))
+          .trigger("keydown", keyCode(key))
+          .then(() => {
+            // eslint-disable-next-line no-unused-expressions
+            expect(callback).to.have.been.called;
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1227]:
```suggestion
          expect(callback).to.have.been.calledOnce;
```

File: src/components/menu/menu.test.js [hunk 11]
```diff
@@ -0,0 +1,1260 @@
...
+      CypressMountWithProviders(
+        <MenuComponentItems onSubmenuClose={callback} />
+      );
+
+      menuItem().eq(0).click();
+      menuComponent(positionOfElement("third"))
+        .click({ multiple: true })
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1242]:
```suggestion
          expect(callback).to.have.been.calledOnce;
```

File: src/components/menu/menu.test.js [hunk 12]
```diff
@@ -0,0 +1,1260 @@
...
+      CypressMountWithProviders(<MenuComponentFullScreen onClose={callback} />);
+
+      cy.viewport(1200, 800);
+      menuItem().eq(positionOfElement("first"), div).click();
+      closeIconButton()
+        .eq(0)
+        .click()
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1256]:
```suggestion
          expect(callback).to.have.been.calledOnce;
```

**Review by DlgSHi — CHANGES_REQUESTED** [review]
I think those TC are missing:
1. Scenario: Check that inner Menu without link is NOT available with tabbing in Fullscreen Menu
    Given I open "Menu Test" component page "menu-full-screen-story"
    When I continue to hit Tab key 8 times
    Then inner menu without active redirection is not focused

2. Scenario: If there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item
    Given I open "Menu Test" component page "long-labels-story"
    When I hover over "first" expandable Menu component
    Then width of submenu should equal the width of the "second" submenu item

3. Scenario: If there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item
    Given I open "Menu Test" component page "long-labels-story"
    When I hover over "second" expandable Menu component
    Then width of submenu should equal the width of "second" expandable Menu component

4. And few props are missing as well:
 - href, 
 - rel, 
 - parent, 
 - parentVariant

File: src/components/menu/menu.test.js [hunk 13]
```diff
@@ -0,0 +1,1260 @@
...
+
+      pressTABKey(1);
+      cy.wait(50);
+      cy.focused().trigger("keydown", keyCode("Enter"));
+      cy.wait(50);
+      pressTABKey(2);
+      cy.wait(50);
+      cy.focused().trigger("keydown", keyCode("uparrow"));
+      searchDefaultInput().should("have.focus");
+    });
```
Comment (ZhuoyuJin) [inline src/components/menu/menu.test.js:435]: Can we combine this test case with the above test case using it.each["downarrow", "uparrow"]?

File: src/components/menu/menu.test.js [hunk 14]
```diff
@@ -0,0 +1,1260 @@
...
+          <MenuSegmentTitle>segment title</MenuSegmentTitle>
+          <MenuItem href="#">Item Submenu Two</MenuItem>
+        </MenuItem>
+      </Menu>
+    </Box>
+  );
+};
+
+context("Testing Menu component", () => {
+  describe("should render Menu component", () => {
```
Comment (ZhuoyuJin) [inline src/components/menu/menu.test.js:260]: could you please change it to something like "check props for Menu component"

File: src/components/menu/menu.test.js [hunk 15]
```diff
@@ -0,0 +1,1260 @@
...
+
+    it.each([
+      ["open", true, "be.visible"],
+      ["closed", false, "not.be.visible"],
+    ])(
+      "should verify that Menu Fullscreen is %s when isOpen prop is %s",
+      (value, boolVal, state) => {
+        CypressMountWithProviders(<MenuComponentFullScreen isOpen={boolVal} />);
+
+        cy.viewport(1200, 800);
```
[the remark this answers is not in the document]
Comment (stephenogorman) [reply src/components/menu/menu.test.js]: Test will fail when boolVal is true if we don't have the viewport.

**Comment by stephenogorman** [comment]
> I think those TC are missing:
> 
>     1. Scenario: Check that inner Menu without link is NOT available with tabbing in Fullscreen Menu
>        Given I open "Menu Test" component page "menu-full-screen-story"
>        When I continue to hit Tab key 8 times
>        Then inner menu without active redirection is not focused
> 
>     2. Scenario: If there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item
>        Given I open "Menu Test" component page "long-labels-story"
>        When I hover over "first" expandable Menu component
>        Then width of submenu should equal the width of the "second" submenu item
> 
>     3. Scenario: If there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item
>        Given I open "Menu Test" component page "long-labels-story"
>        When I hover over "second" expandable Menu component
>        Then width of submenu should equal the width of "second" expandable Menu component
> 
>     4. And few props are missing as well:
> 
> 
>     * href,
> 
>     * rel,
> 
>     * parent,
> 
>     * parentVariant

I added tests for scenario 1 and the props which mere missing. I think scenarios 2 & 3 were actually covered, but the titles weren't clear enough. I have changed those now to:
"should verify if there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item"
&
"should verify if there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item"

**Review by ZhuoyuJin — DISMISSED** [review]

File: src/components/menu/menu.test.js [hunk 16]
```diff
@@ -277,35 +278,24 @@ context("Testing Menu component", () => {
...
-        "have.attr",
-        "data-component",
-        "link"
-      );
-      innerMenu(positionOfElement("third"), span).should(
-        "have.attr",
-        "data-component",
-        "link"
-      );
+      for (let i = 0; i < 4; i++) {
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:287]:
```suggestion
      for (let i = 0; i < position.length; i++) {
```

File: src/components/menu/menu.test.js [hunk 17]
```diff
@@ -1166,6 +996,216 @@ context("Testing Menu component", () => {
...
+      }
+    );
+  });
+
+  describe("check props for Menu Fullscreen component", () => {
+    beforeEach(() => {
+      cy.viewport(1200, 800);
+    });
+
+    const menuClick = (position) => {
```
Comment (DlgSHi) [inline src/components/menu/menu.test.js:1074]: why we can't move this into `beforeEach` as well?

File: src/components/menu/menu.test.js (hunk 8 above)
Comment (DlgSHi) [reply src/components/menu/menu.test.js:978]: not fixed

**Review by DlgSHi — APPROVED** [review]

**Pull request merged**